clippy_lints/src/unnecessary_blocking_ops.rs

```diff
@@ -0,0 +1,173 @@
+    ///
+    /// ### Why is this bad?
+    /// Blocking a thread prevents tasks being swapped, causing other tasks to stop running
+    /// until the thread is no longer blocked, which might not be as expected in an async context.
+    ///
+    /// ### Known problems
+    /// Not all blocking operations can be detected, as for now, this lint only detects a small
+    /// set of functions in standard library by default. And some of the suggestions might need
+    /// additional features to work properly.
+    ///
+    /// ### Example
+    /// ```rust
+    /// use std::time::Duration;
+    /// pub async fn foo() {
+    ///     std::thread::sleep(Duration::from_secs(5));
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```ignore
+    /// use std::time::Duration;
+    /// pub async fn foo() {
+    ///     tokio::time::sleep(Duration::from_secs(5));
+    /// }
+    /// ```
+    #[clippy::version = "1.74.0"]
+    pub UNNECESSARY_BLOCKING_OPS,
+    nursery,
+    "blocking operations in an async context"
+}
+
+pub(crate) struct UnnecessaryBlockingOps {
+    blocking_ops: Vec<String>,
+    blocking_ops_with_suggs: Vec<[String; 2]>,
+    /// Map of resolved funtion def_id with suggestion string after checking crate
+    id_with_suggs: FxHashMap<DefId, Option<String>>,
+    /// Keep track of visited block ids to skip checking the same bodies in `check_body` calls
+    visited_block: HirIdSet,
+}
+
+impl UnnecessaryBlockingOps {
+    pub(crate) fn new(blocking_ops: Vec<String>, blocking_ops_with_suggs: Vec<[String; 2]>) -> Self {
+        Self {
+            blocking_ops,
+            blocking_ops_with_suggs,
+            id_with_suggs: FxHashMap::default(),
+            visited_block: HirIdSet::default(),
+        }
+    }
+}
+
+impl_lint_pass!(UnnecessaryBlockingOps => [UNNECESSARY_BLOCKING_OPS]);
+
+static HARD_CODED_BLOCKING_OPS: [&[&str]; 21] = [
+    &["std", "thread", "sleep"],
+    // Filesystem functions
+    &["std", "fs", "try_exists"],
+    &["std", "fs", "canonicalize"],
+    &["std", "fs", "copy"],
+    &["std", "fs", "create_dir"],
+    &["std", "fs", "create_dir_all"],
+    &["std", "fs", "hard_link"],
+    &["std", "fs", "metadata"],
+    &["std", "fs", "read"],
+    &["std", "fs", "read_dir"],
+    &["std", "fs", "read_link"],
+    &["std", "fs", "read_to_string"],
+    &["std", "fs", "remove_dir"],
+    &["std", "fs", "remove_dir_all"],
+    &["std", "fs", "remove_file"],
+    &["std", "fs", "rename"],
+    &["std", "fs", "set_permissions"],
+    &["std", "fs", "symlink_metadata"],
+    &["std", "fs", "write"],
+    // IO functions
+    &["std", "io", "copy"],
+    &["std", "io", "read_to_string"],
+];
+
+impl<'tcx> LateLintPass<'tcx> for UnnecessaryBlockingOps {
+    fn check_crate(&mut self, cx: &LateContext<'tcx>) {
+        // Avoids processing and storing a long list of paths if this lint was allowed entirely
+        if is_lint_allowed(cx, UNNECESSARY_BLOCKING_OPS, CRATE_HIR_ID) {
```

Comment:
Lints levels can change anywhere.

Reply:
@Jarcho Sorry for late response, I was a bit busy dealing with other stuffs.

I'm not sure if I fully understand this phrase:
> Closures and nested functions inside an async context might run in an async context, but they may also not.

So, if there's an async function like this

```rust
async fn foo() {
    fn bar() {
        // hehe, sleep!
        std::time::sleep(INF)
    }
}
```

Then it shouldn't trigger any warning unless the function `bar` was called in `foo` later, correct?

Reply:
That is correct. Unless the call occurs directly within `foo` there's no way to know if the function is called in an async context. Something like `thread::spawn(|| sleep(10))` or `thread::spawn(bar)` would be fine since the `sleep` call doesn't block the current thread.

Reply:
> That is correct. Unless the call occurs directly within `foo` there's no way to know if the function is called in an async context. Something like `thread::spawn(|| sleep(10))` or `thread::spawn(bar)` would be fine since the `sleep` call doesn't block the current thread.

Ok, I think my code logic already prevent linting on such case, but I'm not sure, I'll add a few test cases tomorrow just in case